src/CityOnPlanet.cpp

```diff
@@ -202,123 +312,345 @@ CityOnPlanet::~CityOnPlanet()
 			// We don't want to create a separate Body for each instance of the buildings, so we
 			// scam the code by pretending we're part of the host planet.
-			Geom *geom = new Geom(cmesh->GetGeomTree(), orientcalc[orient], cent, GetPlanet());
+			Geom *geom = new Geom(cmesh->GetGeomTree(), orientcalc[orient], pos, GetPlanet());
 
 			// add it to the list of buildings to render
-			m_buildings.push_back({ bt.instIndex, float(cmesh->GetRadius()), orient, cent, geom });
+			m_buildings.push_back({ typeIndex, float(cmesh->GetRadius()), orient, pos, geom });
+
 		}
 	}
 
-	Aabb buildAABB;
-	for (std::vector<BuildingDef>::const_iterator iter = m_buildings.begin(), itEND = m_buildings.end(); iter != itEND; ++iter) {
-		buildAABB.Update((*iter).pos - p);
+	_genTimer.Stop();
+
+	Log::Verbose("\tCityOnPlanet: generated {} buildings in {}ms ( {} skipped, {} discarded, {} occupied, {} underwater, {} avg rolls / building )",
+		m_buildings.size(), _genTimer.milliseconds(), skippedCells, discardedCells, occupiedCells, underwaterCells, rarityRolls / double(m_buildings.size()));
+
+	_genTimer.SoftReset();
+
+	// Compute the total AABB of this city
+	Aabb totalAABB;
+	const vector3d &stationOrigin = station->GetPosition();
+
+	for (const auto &buildingInst : m_buildings) {
+		totalAABB.Update(buildingInst.pos - stationOrigin);
+	}
+
+	m_realCentre = totalAABB.min + ((totalAABB.max - totalAABB.min) * 0.5);
+	m_clipRadius = totalAABB.GetRadius();
+
+	_genTimer.Stop();
+
+	Log::Verbose("\tCityOnPlanet: generated AABB for city in {}ms", _genTimer.milliseconds());
+
+	// Release the memory once we're done generating and reset
+	m_gridBitset.reset();
+	m_gridPitch = 0;
+	m_gridLen = 0;
+}
+
+// Calculate the radius for this city based on the SystemBody's parameters
+void CityOnPlanet::CalcCityRadius(const SystemBody *body)
+{
+	double population = body->GetPopulation();
+	double atmo = body->GetAtmosOxidizing();
+
+	for (const CityRadiusDef &radii : reverse_container(m_cityType->sizeDefs)) {
+		bool isLast = (&radii == &m_cityType->sizeDefs.front());
+
+		if (population > radii.population && !isLast)
+			continue;
+
+		m_cityRadius = radii.baseSize + (atmo * radii.atmoSize) + (m_rand.Double() * radii.randomSize);
+		m_cityDensity = radii.density * 0.8;
+		break;
+	}
+}
+
+#ifdef __BIG_ENDIAN__
+#error "CityOnPlanet does not (yet) support big-endian architectures"
+#endif
+
+// Supports up to 32-cell wide buildings to ensure 8-byte aligned addressing in a single operation
+static inline uint64_t CalcBitmaskForGrid(uint32_t x, uint32_t xsize)
+{
+	uint32_t maskOffset = x & CityOnPlanet::CELLMASK;
+	uint64_t bitmask = uint64_t((1 << xsize) - 1) << maskOffset;
+	// note: need crossplatform bswap_64 here for big-endian compat
+	// we should be able to swap the bitmask rather than the in-memory bytes and have the same effect
+
+	return bitmask;
+}
+
+void CityOnPlanet::SetGridOccupancy(uint32_t x, uint32_t y, const uint8_t size[2], bool occupied)
```

Comment:
`bool occupied` is not used.

Reply:
I had originally intended to allow both setting and clearing of grid bits, but never actually used it. A branch inside the inner loop would be counterproductive anyways, so thanks for noticing!